src/katgpucbf/fgpu/engine.py

```diff
@@ -895,6 +914,33 @@ async def _chunk_send_and_cleanup(
                 chunk.cleanup()
                 chunk.cleanup = None  # Potentially helps break reference cycles
 
+    def _update_dig_power_sensors(self, dig_total_power: accel.HostArray | None, out_item: OutQueueItem) -> None:
+        """Update digitiser power sensors.
+
+        Helper function for keeping the complexity of :meth:`run_transmit` down to manageable levels.
+        """
+        if dig_total_power is not None:
+            if out_item.all_present:
+                for pol, trg in enumerate(dig_total_power):
+                    total_power = float(trg)
+                    avg_power = total_power / (out_item.n_spectra * self.output.spectra_samples)
+                    # Normalise relative to full scale. The factor of 2 is because we
+                    # want 1.0 to correspond to a sine wave rather than a square wave.
+                    avg_power /= ((1 << (self.engine.src_layout.sample_bits - 1)) - 1) ** 2 / 2
+                    # If for some reason there's zero power, avoid reporting
+                    # -inf dB by assigning the most negative representable value
+                    avg_power_db = 10 * math.log10(avg_power) if avg_power else np.finfo(np.float64).min
+                    self.engine.sensors[f"input{pol}.dig-rms-dbfs"].set_value(
+                        avg_power_db, timestamp=self.engine.time_converter.adc_to_unix(out_item.next_timestamp)
+                    )
+            else:
+                for pol in range(N_POLS):
+                    self.engine.sensors[f"input{pol}.dig-rms-dbfs"].set_value(
+                        np.finfo(np.float64).min,
+                        status=aiokatcp.Sensor.Status.FAILURE,
+                        timestamp=self.engine.time_converter.adc_to_unix(out_item.next_timestamp),
```

Comment:
You could  make this code slightly more DRY by extracting `self.engine.time_converter.adc_to_unix(out_item.next_timestamp)` into a local variable before the if-else.